src/Crypto/RNG.hs

```diff
@@ -24,118 +20,122 @@
 -- The access to the RNG state is captured by a class.  By making
 -- instances of this class, client code can enjoy RNG generation from
 -- their own monads.
-module Crypto.RNG (
-  -- * CryproRNG class
+module Crypto.RNG
+  ( -- * CryproRNG class
     module Crypto.RNG.Class
-  -- * Generation of strings and numbers
+    -- * Generation of strings and numbers
   , CryptoRNGState
   , newCryptoRNGState
+  , newCryptoRNGStateSized
   , unsafeCryptoRNGState
   , randomBytesIO
-  , randomR
-  -- * Generation of values in other types
-  , Random(..)
-  , boundedIntegralRandom
-  -- * Monad transformer for carrying rng state
+  , randomIO
+  , randomRIO
+    -- * Monad transformer for carrying rng state
   , CryptoRNGT
   , mapCryptoRNGT
   , runCryptoRNGT
   , withCryptoRNGState
   ) where
 
-import Prelude hiding (fail)
 import Control.Applicative
 import Control.Concurrent
 import Control.Monad.Base
-import Control.Monad.Catch hiding (fail)
-import Control.Monad.Cont hiding (fail)
-import Control.Monad.Except hiding (fail)
-import Control.Monad.Fail (MonadFail(..))
-import Control.Monad.Reader hiding (fail)
+import Control.Monad.Catch
+import Control.Monad.Cont
+import Control.Monad.Except
+import Control.Monad.Reader
 import Control.Monad.Trans.Control
 import Crypto.Random
 import Crypto.Random.DRBG
 import Data.Bits
-import Data.ByteString (ByteString, unpack)
-import Data.Int
-import Data.List
-import Data.Word
+import Data.ByteString (ByteString)
+import Data.Either
+import Data.Hashable
+import Data.Primitive.SmallArray
+import qualified Data.ByteString as BS
+import qualified System.Random as R
 
 import Crypto.RNG.Class
 
--- | The random number generator state.  It sits inside an MVar to
--- support concurrent thread access.
-newtype CryptoRNGState = CryptoRNGState (MVar (GenAutoReseed HashDRBG HashDRBG))
+-- | The random number generator state.
+newtype CryptoRNGState = CryptoRNGState (SmallArray (MVar RNG))
+
+-- | Random number generator.
+newtype RNG = RNG (GenBuffered ((GenAutoReseed HashDRBG HashDRBG)))
+
+withRNG :: MVar RNG -> (RNG -> (a, RNG)) -> IO a
+withRNG mrng f = modifyMVar mrng $ \rng -> do
+  (a, newRng) <- pure $ f rng
+  newRng `seq` pure (newRng, a)
+
+-- | Choose the RNG from the pool to use.
+chooseRNG :: CryptoRNGState -> IO (MVar RNG)
+chooseRNG (CryptoRNGState rngs) = liftIO $ do
+  tid <- hash <$> myThreadId
+  pure $ rngs `indexSmallArray` (tid `rem` sizeofSmallArray rngs)
+
+instance R.RandomGen RNG where
+  split = error "split"
+  genWord32 (RNG g) = case genBytes 4 g of
+    Left err       -> error $ "genBytes failed: " ++ show err
+    Right (bs, g') -> (mkWord bs, RNG g')
+  genWord64 (RNG g) = case genBytes 8 g of
+    Left err       -> error $ "genBytes failed: " ++ show err
+    Right (bs, g') -> (mkWord bs, RNG g')
+
+mkWord :: (Bits a, Num a) => ByteString -> a
+mkWord bs = BS.foldl' (\acc w -> shiftL acc 8 .|. fromIntegral w) 0 bs
```

Comment:
This might be totally off-topic, but this piece of code won't work on big endian architectures, like M1 chips. Do we care?

Reply:
AFAIK M1 is little endian, like all ARM processors (they can run in big endian, but almost never are

Reply:
Hmm, but these are random bytes, so their order doesn't really matter. Unless I'm missing something?

Reply:
Yeah, it probably doesn't matter.